File: lib/factorix/cli/commands/launch.rb
# frozen_string_literal: true

module Factorix
  class CLI
    module Commands
      # Launch Factorio game
      #
      # This command launches the Factorio game executable with optional arguments.
      # By default, the game is launched asynchronously (in the background), but certain
      # options like --help and --dump-* are automatically detected and run synchronously.
      class Launch < Base
        require_game_stopped!
        # Game options that require synchronous execution
        #
        # These options output information and exit immediately, so we should
        # wait for them to complete rather than running them in the background.
        SYNCHRONOUS_OPTIONS = %w[
          --dump-data
          --dump-icon-sprites
          --dump-prototype-locale
          --help
          --version
        ].freeze
        private_constant :SYNCHRONOUS_OPTIONS

        # @!parse
        #   # @return [Runtime::Base]
        #   attr_reader :runtime
        #   # @return [Dry::Logger::Dispatcher]
        #   attr_reader :logger
        include Import[:runtime, :logger]

        desc "Launch Factorio game"

        example [
          "                              # Launch Factorio",
          "-- --help                     # Show Factorio help",
          "-- --benchmark save.zip       # Run benchmark"
        ]

        option :wait, type: :flag, default: false, aliases: ["-w"], desc: "Wait for the game to finish"

        # Execute the launch command
        #
        # @param wait [Boolean] whether to wait for the game to finish
        # @param args [Array<String>] additional arguments to pass to Factorio
        # @return [void]
        def call(wait: false, args: [], **)
          logger.info("Launching Factorio", args:)

          async = args.none? {|arg| SYNCHRONOUS_OPTIONS.include?(arg) }

          runtime.launch(*args, async:)
          logger.info("Factorio launched successfully", async:)

          return unless async && wait

          logger.debug("Waiting for game to start")
          wait_while { !runtime.running? }
          logger.debug("Game started, waiting for termination")
          wait_while { runtime.running? }
          logger.info("Game terminated")
        end

        # Wait while a condition is true
        #
        # @yield the condition to check
        # @return [void]
        private def wait_while
          loop do
            break unless yield

            sleep 1
          end
        end
      end
    end
  end
end

File: internal/cli/launch.go
package cli

import (
	"slices"
	"time"

	"github.com/spf13/cobra"
)

// Game options that print and exit without daemonizing; the launch waits
// for these instead of detaching.
var synchronousLaunchOptions = []string{
	"--dump-data",
	"--dump-icon-sprites",
	"--dump-prototype-locale",
	"--help",
	"--version",
}

// launchPollInterval is how often --wait polls the lock file; a variable so
// tests can shorten it.
var launchPollInterval = time.Second

func newLaunchCommand(c *cli) *cobra.Command {
	var wait bool

	cmd := &cobra.Command{
		Use:   "launch [-- <factorio-arg>...]",
		Short: "Launch Factorio game",
		Args:  cobra.ArbitraryArgs,
		RunE: func(cmd *cobra.Command, args []string) error {
			application, err := c.App()
			if err != nil {
				return err
			}
			if err := application.RequireGameStopped(); err != nil {
				return err
			}

			async := !slices.ContainsFunc(args, func(arg string) bool {
				return slices.Contains(synchronousLaunchOptions, arg)
			})
			application.Logger.Info("Launching Factorio", "args", args)
			if err := application.Runtime.Launch(args, async); err != nil {
				return err
			}
			application.Logger.Info("Factorio launched successfully", "async", async)

			if !async || !wait {
				return nil
			}

			// Factorio double-forks, so the started process cannot be
			// waited on; poll the lock file instead: first until the game
			// has created it, then until it disappears.
			application.Logger.Debug("Waiting for game to start")
			if err := waitForRunningState(application.Runtime.IsRunning, true); err != nil {
				return err
			}
			application.Logger.Debug("Game started, waiting for termination")
			if err := waitForRunningState(application.Runtime.IsRunning, false); err != nil {
				return err
			}
			application.Logger.Info("Game terminated")
			return nil
		},
	}
	cmd.Flags().BoolVarP(&wait, "wait", "w", false, "Wait for the game to finish")
	return cmd
}

// waitForRunningState polls isRunning until it reports want.
func waitForRunningState(isRunning func() (bool, error), want bool) error {
	for {
		running, err := isRunning()
		if err != nil {
			return err
		}
		if running == want {
			return nil
		}
		time.Sleep(launchPollInterval)
	}
}
